File: backend/app/repositories/party_repository.py
from typing import List, Dict, Any, Optional
from app.utils.dynamodb import DynamoDBClient
from boto3.dynamodb.conditions import Key

class PartyRepository:
    def __init__(self, table_name: str = "knowyourmla_political_parties", client: DynamoDBClient = None):
        if client:
            self.client = client
        else:
            self.client = DynamoDBClient(table_name=table_name)

    def get_all_parties(self) -> List[Dict[str, Any]]:
        """
        Fetches all party metadata from knowyourmla_political_parties.
        Uses scan to get all items.
        """
        return self.client.scan()

    def get_party_by_id(self, party_id: str) -> Optional[Dict[str, Any]]:
        """
        Fetches party metadata by its PK (e.g., PARTY#dravidamunnetrakazhagam).
        """
        if not party_id.startswith("PARTY#"):
            party_id = f"PARTY#{party_id}"
            
        return self.client.get_item(Key={'PK': party_id, 'SK': 'METADATA'})

File: frontend/lib/repositories/party.repository.ts
import { DynamoDBWrapper } from "../dynamodb";

export class PartyRepository {
  private client: DynamoDBWrapper;

  constructor(tableName: string = process.env.PARTIES_TABLE || "knowyourmla_political_parties") {
    this.client = new DynamoDBWrapper(tableName);
  }

  /**
   * Fetches all political parties (METADATA) using MetadataIndex.
   */
  async getAllParties() {
    return this.client.query({
      IndexName: "MetadataIndex",
      KeyConditionExpression: "SK = :sk",
      ExpressionAttributeValues: { ":sk": "METADATA" },
      ProjectionExpression: "PK, #name, short_name, logo_url, color_bg, color_text, color_border, vote_share",
      ExpressionAttributeNames: {
        "#name": "name"
      }
    });
  }

  async getPartyById(partyId: string) {
    let pk = partyId;
    if (!pk.startsWith("PARTY#")) {
      pk = `PARTY#${pk}`;
    }

    try {
      return this.client.get({
        Key: { PK: pk, SK: "METADATA" },
      });
    } catch (error) {
      console.error(`Error fetching party ${pk}:`, error);
      return null;
    }
  }

  /**
   * Attempts to resolve a party by its slug using direct ID lookups.
   * Returns null if it cannot be found directly.
   */
  async getPartyBySlugDirect(slug: string) {
    const normalizedSlug = slug.toLowerCase().replace(/[^a-z0-9]/g, "");
    
    // 1. Try direct PK lookup with normalized slug
    const direct = await this.getPartyById(`PARTY#${normalizedSlug}`);
    if (direct) return direct;

    // 2. Try exact uppercase short name (common PK pattern: PARTY#DMK)
    const upperShort = await this.getPartyById(`PARTY#${slug.toUpperCase()}`);
    if (upperShort) return upperShort;

    return null;
  }
}
